### PR TITLE
Support Array<jbyte> <-> std::string conversions

### PR DESCRIPTION
**Pull request by spinorx**
Make jni fields/methods copyable.

Array<jbyte> <-> std::string conversions are needed to push protocol buffers
across jni boundaries by serializing and deserializing.

Copyable Method/Fields are needed so that these can be stored in global statics.

### CODE REVIEW

**Comment by jfirebaugh** [comment]
Hi @spinorx, thanks for the contribution! Adding `Array<jbyte>` <-> `std::string` conversions seems reasonable. Could you please move that into a separate commit?

As for the changes to Method/Fields, I'm not sure these are necessary. These classes are already copyable -- that's required for [`Class::GetField`](https://github.com/mapbox/jni.hpp/blob/master/include/jni/class.hpp#L116) and so on to be able to return them by value. Was there a particular issue you were trying to solve by making that change?

**Comment by spinorx** [comment]
Now it has only the std::string <-> jniArray<jbyte> changes.  Will put separate diff for copyable along with error I see.

**Comment by spinorx** [comment]
And Thanks for the library,  was helpful!

File: include/jni/arraylike.hpp [hunk 1]
```diff
@@ -30,8 +30,11 @@ namespace jni
     template < class E, std::size_t n >    E const * ArraylikeData(const std::array<E,n>& a)              { return a.data(); }
     template < class E, class A >          E       * ArraylikeData(      std::vector<E,A>& a)             { return a.data(); }
     template < class E, class A >          E const * ArraylikeData(const std::vector<E,A>& a)             { return a.data(); }
-    template < class C, class T, class A > C       * ArraylikeData(      std::basic_string<C,T,A>& a)     { return &a[0]; }
-    template < class C, class T, class A > C const * ArraylikeData(const std::basic_string<C,T,A>& a)     { return &a[0]; }
+    template < class C, class T, class A > C       * ArraylikeData(      std::basic_string<C,T,A>& a)     { return a.data(); }
```
Comment (jfirebaugh) [inline include/jni/arraylike.hpp]: The reason to use `&a[0]` as opposed to `a.data()` is that [the non-`const` overload for `data()`](http://en.cppreference.com/w/cpp/string/basic_string/data) is provided only in C++17, and this library maintains compatibility with C++14. Could you revert this change please?

File: include/jni/arraylike.hpp [hunk 2]
```diff
@@ -30,8 +30,11 @@ namespace jni
...
+    template < class C, class T, class A > C const * ArraylikeData(const std::basic_string<C,T,A>& a)     { return a.data(); }
+    // For supporting std::string <-> jbyte conversions
+    template < class T, class A >      jbyte       * ArraylikeData(      std::basic_string<char,T,A>& a)     { return (jbyte*)a.data(); }
```
Comment (jfirebaugh) [inline include/jni/arraylike.hpp]: These overloads overlap with the two templated overloads `ArraylikeData` above. Since they're non-templated, they'll take precedence: if you pass `ArraylikeData` a `std::string`, the compiler will select the `jbyte*`-returning version, which is not always what you want. So instead of adding new `ArraylikeData` overloads, I think you'll need to adjust the definition of the added `MakeAnything` overloads to include the cast. Also, it should be a `reinterpret_cast` rather than a C-style cast. The latter are prohibited in jni.hpp. (You can run `make test` to make sure your additions compile.)

File: include/jni/arraylike.hpp (hunk 1 above)
Comment (spinorx) [reply include/jni/arraylike.hpp]: a[0] will not work with empty strings though :(  Wouldn't it be better to condition it on C++17?

Comment (jfirebaugh) [reply include/jni/arraylike.hpp]: It does work (since C++11): it's defined to return a reference to the character with value CharT() (the null character).

File: include/jni/arraylike.hpp (hunk 2 above)
Comment (spinorx) [reply include/jni/arraylike.hpp]: Makes sense.

**Comment by spinorx** [comment]
done ptal.
